utopia-remix/app/routes/v1.project._index.tsx

```diff
@@ -0,0 +1,104 @@
+import { Status } from '../util/statusCodes'
+import { ApiError } from '../util/errors'
+import { action as createProjectIdInternal } from './v1.projectid'
+import { action as createProjectInternal } from './v1.project.$id'
+
+export async function loader(args: LoaderFunctionArgs) {
+  return handle(args, {
+    OPTIONS: handleOptions,
+  })
+}
+
+export async function action(args: ActionFunctionArgs) {
+  return handle(args, {
+    PUT: {
+      handler: createNewProject,
+      validator: ALLOW,
+    },
+  })
+}
+
+async function createNewProject(req: Request, params: Params<string>) {
+  // get a new project id
+  const projectIdRequest = new Request(withPath(req, '/v1/projectid'), {
+    method: 'POST',
+    headers: req.headers,
+  })
+  const projectIdResponse = (await createProjectId(projectIdRequest)) as { id: string }
+  const { id } = projectIdResponse
+
+  // prepare data for creating the project
+  const requestData = (await req.json()) as { content: object; name: string; accessLevel: string }
+  const { content, name, accessLevel: accessLevelParam } = requestData
+  const headers = new Headers(req.headers)
+  headers.delete('content-length')
+  const createProjectRequest = new Request(withPath(req, `/v1/project/${id}`), {
+    method: 'PUT',
+    headers: headers,
+    body: JSON.stringify({ content: content, name: name }),
+  })
+
+  // create the project
+  const project = (await createProject(createProjectRequest, { id: id })) as {
+    ownerId: string | null
+  }
+
+  // handle access level setting
+  const accessLevel: AccessLevel =
+    accessLevelParamToAccessLevel(accessLevelParam) ?? AccessLevel.PRIVATE
+  await createProjectAccess({
+    projectId: id,
+    accessLevel: accessLevel,
+    creatorId: project.ownerId ?? null,
+  })
+
+  return project
+}
+
+function accessLevelParamToAccessLevel(accessLevelParam: string | null): AccessLevel | null {
+  if (accessLevelParam == null || accessLevelParam == '') {
+    return null
+  }
+  switch (accessLevelParam) {
+    case 'private':
+      return AccessLevel.PRIVATE
+    case 'public':
+      return AccessLevel.PUBLIC
+    case 'collaborative':
+      return AccessLevel.COLLABORATIVE
+    default:
+      throw new ApiError('Invalid access level', Status.BAD_REQUEST)
+  }
+}
+
+function withPath(req: Request, path: string): URL {
+  const url = new URL(req.url)
+  url.pathname = path
+  return url
+}
+
+async function createProjectId(req: Request) {
+  // const response = await fetch(req)
+  const response = (await createProjectIdInternal({
+    request: req,
+    params: {},
+    context: {},
+  })) as Response
+  return await response.json()
+}
+
+async function createProject(req: Request, params: Params<string>) {
+  // const response = await fetch(req)
+  const response = (await createProjectInternal({
+    request: req,
+    params: params,
+    context: {},
+  })) as Response
+  return await response.json()
+}
```

Comment:
These two functions are equivalent, I don't think there's a need to have 2, just one helper that does the given request and returns its json response (with a generic function name)

Reply:
I think that calling the actions directly is not great – I'd suggest to either call the handlers themselves, or do a proper HTTP call with `fetch`.

Reply:
actually since we'll be probably removing `projectid` and `PUT /project/{id}` in the near future, maybe it's better to even call `proxy` directly here. changed.